packages/api-client-core/src/support.ts

```diff
@@ -495,3 +495,193 @@ export const storageAvailable = (type: "localStorage" | "sessionStorage") => {
     return false;
   }
 };
+
+export const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+/**
+ * A FIFO queue that executes functions in order, waiting for each to finish before executing the next.
+ */
+export class PQueue {
+  /**
+   * The number of functions waiting to be executed.
+   */
+  get length(): number {
+    return this.queue.length;
+  }
+
+  /**
+   * A queue of functions waiting to be executed.
+   */
+  private queue: (() => unknown | Promise<unknown>)[] = [];
+
+  /**
+   * Whether the queue is currently being dequeued.
+   */
+  private isDequeuing = false;
+
+  /**
+   * Dequeue and execute all the functions in the queue in order.
+   */
+  private dequeue(): void {
+    if (this.isDequeuing) {
+      return;
+    }
+
+    void (async () => {
+      this.isDequeuing = true;
+
+      let fn = this.queue.shift();
+      while (fn) {
+        try {
+          await fn();
+        } catch {
+          // swallow errors to avoid unhandled promise rejections
+        }
+        fn = this.queue.shift();
+      }
+
+      this.isDequeuing = false;
+    })();
```

Comment:
The biggest reason I ended up doing `try/catch` is to catch errors from both sync and async functions. 

The `queue` should only contain `async functions` but in the rare case it didn't, `fn().catch(…)` would throw because `fn()` wouldn't return a promise.

Reply:
[`f094575` (#256)](https://github.com/gadget-inc/js-clients/pull/256/commits/f09457521136ea86564496e0ae368cec67ebfafd) 👍